src/python/pants/option/global_options.py

```diff
@@ -278,6 +290,7 @@ def from_options(
         full_options: Options,
         env: CompleteEnvironment,
         prior_result: AuthPluginResult | None = None,
+        remote_auth_plugin_func: Callable | None = None,
```

Comment:
Having a fully specified `Protocol` for this type would be good.

Reply:
I tried to do it, but this feels like a rabbit hole which is out of scope for this PR.
since arguments are named, if I do: 
```python
AuthPluginFunctionProtocol = Callable[[Dict[str, str], Dict[str, str], Options, Mapping[str, str], Union[AuthPluginResult, None]],   AuthPluginResult]```
and use that instead of callable, mypy complains:
```
src/python/pants/option/global_options.py:391:30: error: Unexpected keyword
argument "initial_execution_headers"  [call-arg]
            auth_plugin_result = remote_auth_plugin_func(
                                 ^
src/python/pants/option/global_options.py:391:30: error: Unexpected keyword
argument "initial_store_headers"  [call-arg]
            auth_plugin_result = remote_auth_plugin_func(
                                 ^
src/python/pants/option/global_options.py:391:30: error: Unexpected keyword
argument "options"  [call-arg]
            auth_plugin_result = remote_auth_plugin_func(
                                 ^
src/python/pants/option/global_options.py:391:30: error: Unexpected keyword
argument "env"  [call-arg]
            auth_plugin_result = remote_auth_plugin_func(
                                 ^
src/python/pants/option/global_options.py:391:30: error: Unexpected keyword
argument "prior_result"  [call-arg]
            auth_plugin_result = remote_auth_plugin_func(
                                 ^
Found 5 errors in 1 file (checked 29 source files)
```

so this requires usage of [NamedArg](https://mypy.readthedocs.io/en/stable/additional_features.html#extended-callable-types) which is available only in [mypy_extensions](https://pypi.org/project/mypy-extensions/) currently not used in the pants repo.

beside all that, since this is runtime behavior thing, any type checks here are somewhat pointless since the auth plugin function is loaded at runtime and any validation mypy does in the pants repo itself feels pointless and I don't see what kind of errors the additional and more accurate typing information will prevent.

Reply:
Sorry, I meant literally `Protocol`: https://mypy.readthedocs.io/en/stable/protocols.html#simple-user-defined-protocols

Reply:
Not sure how to use here since the function is implemented as a function and not a class and from a quick read of the docs you linked to, it seems that Protocol is intended to be use in classes based interfaces.

At any case, as I said, this still feels out of scope for this PR (we don't have it with the current implementation)  and I don't see the benefit (since this is a runtime behavior) what kind of bugs do you think this kind of annotation will prevent ?

Reply:
To make a `Protocol` for a `Callable` you define the `__call__` method: an example: https://github.com/pantsbuild/pants/blob/258ab2602a05b1450540d8d3d89f51407aa416fa/src/python/pants/pantsd/pants_daemon_core.py#L28-L34

But yea, as you said: we don't define this API yet, so I'm fine with it being out of scope.

Reply:
got it, I'll have a follow up PR for this.
since if we this type is exposed from pants, then the plugin implementation can use it.